### PR TITLE
INT: Filter import candidates

### PR DESCRIPTION
**Pull request by Undin**
part of #1737

* drop useless candidates from stdlib
* don't add import candidate if there is import candidate with reexport from ancestor module

### CODE REVIEW

**Comment by Undin** [comment]
@vlad20012 r?

File: src/main/kotlin/org/rust/ide/intentions/import/ImportNameIntention.kt [hunk 1]
```diff
@@ -151,6 +152,56 @@ class ImportNameIntention : RsElementBaseIntentionAction<ImportNameIntention.Con
         return !(element.parent is RsMembers && element.ancestorStrict<RsTraitItem>() != null)
     }
 
+    private fun filterUselessImports(candidates: List<ImportCandidate>, fileAttributes: RsFile.Attributes): List<ImportCandidate> {
+        val finalCandidates = mutableListOf<ImportCandidate>()
+        val stdlibCandidates = mutableListOf<ImportCandidate>()
+        val otherCandidates = mutableListOf<ImportCandidate>()
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/intentions/import/ImportNameIntention.kt]: Why we need `otherCandidates` here? We can just add "other candidates" directly to `finalCandidates`

File: src/main/kotlin/org/rust/ide/intentions/import/ImportNameIntention.kt [hunk 2]
```diff
@@ -152,38 +152,33 @@ class ImportNameIntention : RsElementBaseIntentionAction<ImportNameIntention.Con
...
+            val container = if (pkg.origin == PackageOrigin.STDLIB) stdlibCandidates else candidatesWithPackage
+            container += candidate to pkg
         }
 
-        finalCandidates += filterStdlibCandidates(stdlibCandidates, fileAttributes)
-        finalCandidates += otherCandidates
-        return finalCandidates
+        candidatesWithPackage += filterStdlibCandidates(stdlibCandidates, fileAttributes)
+        val pkgToCandidates = candidatesWithPackage.groupBy({ (_, pkg) -> pkg }, { (candidate, _) -> candidate })
+        return pkgToCandidates.flatMapTo(ArrayList()) { (_, candidates) -> filterInPackage(candidates) }
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/intentions/import/ImportNameIntention.kt]: `ArrayList()`?

File: src/main/kotlin/org/rust/ide/intentions/import/ImportNameIntention.kt [hunk 3]
```diff
@@ -130,14 +133,95 @@ class ImportNameIntention : RsElementBaseIntentionAction<ImportNameIntention.Con
...
+
+        for (candidate in candidates) {
+            val pkg = candidate.importItem.containingCargoTarget?.pkg ?: continue
+            val container = if (pkg.origin == PackageOrigin.STDLIB) stdlibCandidates else candidatesWithPackage
+            container += candidate to pkg
+        }
+
+        candidatesWithPackage += filterStdlibCandidates(stdlibCandidates, fileAttributes)
+        val pkgToCandidates = candidatesWithPackage.groupBy({ (_, pkg) -> pkg }, { (candidate, _) -> candidate })
+        return pkgToCandidates.flatMapTo(ArrayList()) { (_, candidates) -> filterInPackage(candidates) }
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/intentions/import/ImportNameIntention.kt]: `ArrayList()`?

Comment (Undin) [reply src/main/kotlin/org/rust/ide/intentions/import/ImportNameIntention.kt]: Hm, it's an artifact from some intermediate version.
Of course, just `flatMap` need here

**Comment by Undin** [comment]
bors r=vlad20012

**Pull request merged**